maven_iuvs/echelle.py

```diff
@@ -825,3 +966,1289 @@ def find_files_with_geometry(file_index):
+    """
+    Produces a spectrum averaged along the spatial dimension of the slit 
+    NOTE: This is called by both the l1a-->l1c pipeline and the quicklook maker,
+    so don't get rid of the coadded functionality and think you're clever; we need
+    that for the quicklooks.
+    NOTE: Although this function could operate on any detector image and produce
+    some kind of result, it is only really relevant to the orders containing the 
+    Lyman alpha emissions.
+
+    Parameters:
+    ----------
+    data : array
+           3D numpy array of image detector data, dark subtracted and cleaned. 
+    light_fits : astropy.io.fits instance
+                 File with light observation
+    average : boolean
+              whether to get the average spectrum across the slit, False by default.
+    coadded : boolean
+              If True, will get a spectrum averaged across integrations. If False, 
+              will only get the spectrum for one detector frame at a time.
+    integration : int
+                  Integration frame to use for the specrum. Used if coadded=False.
+    clean_data : boolean 
+                 whether to perform the data cleaning routines
+
+    Returns:
+    ----------
+    spectrum : array
+               Spectrum in total DN summed over the spatial dimension
+    
+    """
+    # Collect pixel range which we need to find the slit start and end 
+    si1, si2 = get_ech_slit_indices(light_fits)
+
+    if coadded:
+        spectrum = np.sum(data[si1:si2+1, :], axis=0) # +1 because of the way python does indices.
+    else:
+        # Sum up the spectra over the range in which Ly alpha is visible on the slit (not outside it)
+        # This spectrum is thus in DN
+        spectrum = np.sum(data[integration, si1:si2+1, :], axis=0)
+        
+    if average:
+        spectrum = spectrum / (si2 - si1)
+
+    return spectrum
+
+
+def add_in_quadrature(uncertainties, light_fits, integration=0): 
+    """
+    Similar to get_spectrum, but adds up the uncertainties in what is hopefully 
+    the correct manner. 
+
+    Parameters:
+    ----------
+    data : array
+           3D numpy array of image detector data, dark subtracted and cleaned. 
+    light_fits : astropy.io.fits instance
+                 File with light observation
+    integration : int
+                  Integration frame to use for the specrum. Used if coadded=False.
+    clean_data : boolean 
+                 whether to perform the data cleaning routines
+
+    Returns:
+    ----------
+    spectrum : array
+               Spectrum in total DN summed over the spatial dimension
+    
+    """
+    # Collect pixel range which we need to find the slit start and end 
+    si1, si2 = get_ech_slit_indices(light_fits)
+
+    total_uncert = np.sqrt( np.sum( (uncertainties[integration, si1:si2+1, :])**2, axis=0) )
+    # +1 because of the way python does indices.
+    return total_uncert
+
+
+def dx_array(x):
```

Comment:
this duplicates some aspects of the code used to get wavelength bin edges, consider combining?

Reply:
Yeah, this is a relic

Reply:
Fixed by: f0c60d192fd4cfebf083e2c2a9930d1f48e90bcf